src/timetable_gui.cpp

```diff
@@ -438,32 +437,36 @@ struct TimetableWindow : Window {
 				bool show_late = this->show_expected && v->lateness_counter > DAY_TICKS;
 				Ticks offset = show_late ? 0 : -v->lateness_counter;
 
-				bool rtl = _current_text_dir == TD_RTL;
-				Rect abbr = tr.WithWidth(this->deparr_abbr_width, rtl);
-				Rect time = tr.WithWidth(this->deparr_time_width, !rtl);
-
 				for (int i = this->vscroll->GetPosition(); i / 2 < v->GetNumOrders(); ++i) { // note: i is also incremented in the loop
 					/* Don't draw anything if it extends past the end of the window. */
 					if (!this->vscroll->IsVisible(i)) break;
 
 					if (i % 2 == 0) {
+						/* Draw an arrival time. */
 						if (arr_dep[i / 2].arrival != INVALID_TICKS) {
-							DrawString(abbr.left, abbr.right, tr.top, STR_TIMETABLE_ARRIVAL_ABBREVIATION, i == selected ? TC_WHITE : TC_BLACK);
+							/* First set the offset and text colour based on the expected/scheduled mode and some other things. */
+							Ticks this_offset;
+							TextColour colour;
 							if (this->show_expected && i / 2 == earlyID) {
-								SetDParam(0, _date + arr_dep[i / 2].arrival / DAY_TICKS);
-								DrawString(time.left, time.right, tr.top, STR_JUST_DATE_TINY, TC_GREEN);
+								/* Show expected arrival. */
+								this_offset = 0;
+								colour = TC_GREEN;
 							} else {
-								SetDParam(0, _date + (arr_dep[i / 2].arrival + offset) / DAY_TICKS);
-								DrawString(time.left, time.right, tr.top, STR_JUST_DATE_TINY,
-										show_late ? TC_RED : i == selected ? TC_WHITE : TC_BLACK);
+								/* Show scheduled arrival. */
+								this_offset = offset;
+								colour = show_late ? TC_RED : (i == selected ? TC_WHITE : TC_BLACK);
```

Comment:
If I follow correctly, with separated abbreviation the selected state was always visible on the abbreviated part.
Now if the string is green you'll never see the selected state, same for late as the string will be full red.

Reply:
Ah, yes, that's correct. I might have to return to drawing the abbreviation as a separate string, as I don't know a way to draw two colours ({PUSH/POP_COLOUR} only works with colour string codes provided in the string, I think).

Reply:
Mildly unrelated, but since you're in the area, could do with splitting this switch into multiple functions. The indentation level in this function is getting a bit extreme

Reply:
Indeed, I'll see what I can do. 🙂

Reply:
> Ah, yes, that's correct. I might have to return to drawing the abbreviation as a separate string, as I don't know a way to draw two colours ({PUSH/POP_COLOUR} only works with colour string codes provided in the string, I think).

It might be possible to introduce a new `{COLOUR}` command taking a parameter in the `SCC_BLUE`-`SCC_BLACK` range (or 0-16 range which might be better for GSText).
I see exactly how to do it, so I may try.